lazy parsing of iterations

franzpoeschel opened the pull request:
Opening a Series with many iterations is currently a time-intensive procedure due to eager parsing.
This adds an option to lazily parse a Series. If that option is chosen, `Iteration::open()` must be called before accessing. `ReadIterations` will do that automatically.

I previously attempted parsing iterations automatically upon accessing them via the `Container` interface, but there were too many edge cases and it's probably better to do such things explicitly in parallel situations.

TODO:
- [x] It's probably time to use a builder pattern for the Series or something alike, the number of optional parameters keeps growing.
- [x] Python bindings
- [x] Documentation
- [x] Cleanup, ~~including removal of failed first attempt~~
- [x] Support lazy parsing for group-based layout
- [x] Merge #941 first.
- [x] This PR is affected by the issue described in #953. Worked around it in here by using `auxiliary::Option` instead of NaNs to represent no scaling.

franzpoeschel committed:
Lazy parsing

First attempt at an implementation of lazy parsing

todo:
* file-based layout
* iterator access

Basic file-based deferred parsing

Remove changes to Container class

Read eagerly by default

Read iteration upon Iteration::open

Expose to frontend and add some tests

Use Builder Pattern for Series

franzpoeschel committed:
Fix CI error

franzpoeschel committed:
Dead code removal, some commenting

franzpoeschel committed:
Python bindings and tests

ax3l commented on test/SerialIOTest.cpp:
```diff
@@ -59,9 +59,15 @@ write_and_read_many_iterations( std::string const & ext ) {
-        Series read(filename, Access::READ_ONLY);
-        for (auto iteration : read.iterations) {
-            // std::cout << "Reading iteration " << iteration.first << std::endl;
+        Series read = SeriesBuilder()
```
C++ builder pattern.

franzpoeschel committed:
Add explicit test for lazy parsing

franzpoeschel committed:
Do not automatically open an Iteration

franzpoeschel committed:
Group-based layout: Parse only after beginning steps

franzpoeschel committed:
Fix bug when opening a filebased series as groupbased

In that case, we must not attempt to piece back together the filename by
filebased methods, but instead directly use the filename specified by
the user.

franzpoeschel committed:
Document SeriesBuilder

franzpoeschel committed:
Document parseLazily option of the Series constructor

franzpoeschel committed:
Cleanup

ax3l commented on include/openPMD/SeriesBuilder.hpp:
```diff
@@ -0,0 +1,94 @@
+#if openPMD_HAVE_MPI
+    /**
+     * @brief As in the Series constructor. Default is to construct a
+     *        non-parallel Series.Mit
```
Is `Series.Mit` a typo? :)

franzpoeschel replied:
Guess I typed something in the wrong window heh
I've removed the builder thing from this PR now anyway though

ax3l reviewed (commented):
added inline :)

ax3l commented on include/openPMD/Iteration.hpp:
```diff
@@ -153,22 +154,48 @@ class Iteration : public LegacyAttributable
 private:
     Iteration();
 
+    struct DeferredRead
```
Let's add a doxygen string here, just to clarify what this switch does.
Would `DeferredReadAccess` be more explicit? Not sure, ... maybe.

franzpoeschel replied:
If we rename this one anyway, I'd go for `DeferredParseAccess` to avoid confusion with deferred load_chunks

ax3l reviewed (commented):
The implementation of the lazy/deferred parsing reads well. One general comment, I would unify the naming for this in the code base, at the moment we use:
- lazy
- deferred read
- deferred access
- (not) yet accessed

To describe the same thing, which can be confusing as the same concept in the code is meant.

Should we skip the introduction of the term "lazy" and instead use "deferred (iteration) access" in all places?

---

Do we want to make this a Series constructor parameter or JSON option?

---

Should we introduce the builder pattern in a separate PR and/or use it more consistently throughout the code base & examples?

franzpoeschel committed:
Remove SeriesBuilder

franzpoeschel committed:
Make parse_lazily a JSON option

franzpoeschel committed:
Rename things, common naming scheme

parse_lazily -> defer_iteration_parsing
NotYetAccessed -> ParseAccessDeferred
Iteration::deferRead -> Iteration::deferParseAccess
DeferredRead -> DeferredParseAccess
m_deferredRead -> m_deferredParseAccess
parseLazily -> runDeferredParseAccess

franzpoeschel reviewed (commented):
> The implementation of the lazy/deferred parsing reads well. One general comment, I would unify the naming for this in the code base, at the moment we use:
> 
>     * lazy
> 
>     * deferred read
> 
>     * deferred access
> 
>     * (not) yet accessed
> 
> 
> To describe the same thing, which can be confusing as the same concept in the code is meant.
> 
> Should we skip the introduction of the term "lazy" and instead use "deferred (iteration) access" in all places?

I've renamed things now:
```
parse_lazily -> defer_iteration_parsing
NotYetAccessed -> ParseAccessDeferred
Iteration::deferRead -> Iteration::deferParseAccess
DeferredRead -> DeferredParseAccess
m_deferredRead -> m_deferredParseAccess
parseLazily -> runDeferredParseAccess
```
> 
> Do we want to make this a Series constructor parameter or JSON option?

Good idea, done
> 
> Should we introduce the builder pattern in a separate PR and/or use it more consistently throughout the code base & examples?

I introduced the builder pattern since it made for a better way to have a constructor with more than one defaulted argument. Since we're now back at one defaulted argument, we don't necessarily need it anymore. I've isolated the builder pattern to a branch `franzpoeschel/topic-builder` now, should I do a PR?

Also, we can definitely use it more in the examples and tests. In the code base itself probably not so much since we seldomly create a `Series` object ourselves.

ax3l commented on docs/source/details/backendconfig.rst:
```diff
@@ -33,6 +33,18 @@ For a consistent user interface, backends shall follow the following rules:
+Especially when a Series has many iterations, this can be a costly operation and users may wish to defer parsing of iterations to a later point adding ``{"defer_iteration_parsing": true}`` to their JSON configuration.
+
+When parsing non-eagerly, each iteration needs to be explicitly opened with ``Iteration::open()`` before accessing.
+(Notice that Iteration::open() is generally recommended to be used in parallel contexts to avoid parallel file accessing hazards).
```
```suggestion
(Notice that ``Iteration::open()`` is generally recommended to be used in parallel contexts to avoid parallel file accessing hazards).
```

ax3l commented on docs/source/details/backendconfig.rst:
```diff
@@ -33,6 +33,18 @@ For a consistent user interface, backends shall follow the following rules:
+
+When parsing non-eagerly, each iteration needs to be explicitly opened with ``Iteration::open()`` before accessing.
+(Notice that Iteration::open() is generally recommended to be used in parallel contexts to avoid parallel file accessing hazards).
+Using the Streaming API (i.e. SeriesImpl::readIteration()) will do this automatically.
```
```suggestion
Using the Streaming API (i.e. ``SeriesImpl::readIteration()``) will do this automatically.
```

ax3l commented on docs/source/details/backendconfig.rst:
```diff
@@ -33,6 +33,18 @@ For a consistent user interface, backends shall follow the following rules:
+(Notice that Iteration::open() is generally recommended to be used in parallel contexts to avoid parallel file accessing hazards).
+Using the Streaming API (i.e. SeriesImpl::readIteration()) will do this automatically.
+Parsing eagerly might be very expensive for a Series with many iterations, but will avoid bugs by forgotten calls to ``Iteration::open()``.
+
```
Let's add this if true (we should also add CI for this if true):
```suggestion
In complex environments, calling ``Iteration::open()`` on an already open environment does no harm (and does not incur additional runtime cost for additional ``open()`` calls).

```

ax3l commented on include/openPMD/Series.hpp:
```diff
@@ -328,6 +331,9 @@ class SeriesImpl : public AttributableImpl
     std::unique_ptr< ParsedInput > parseInput(std::string);
+    // use a template in order not to expose nlohmann_json to users
+    template< typename JSON >
+    void parseJsonOptions( JSON const & );
```
Since you only call this in the `.cpp` file, you can define this helper function only in the `Series.cpp` file and in a `namespace {}` (as you already do) and avoid that it gets exposed as a function and symbol at all :)

franzpoeschel replied:
It's not that easy because that function accesses private members. But we can theoretically make all members of `SeriesData` public and inherit it privately in `SeriesInternal`. So, the data struct can be passed to non-friend functions for manipulation?
Edit: Pushed a commit that does just that.
Edit: Not so easy either, this [breaks dynamic casting](https://stackoverflow.com/questions/62430684/why-c-requires-public-inheritance-ignoring-friend-declarations-to-make-dynam). I've made it a public inheritance now, meaning that `SeriesInternal` has all members public now. `Series` still hides them.

ax3l commented on test/python/unittest/API/APITest.py:
```diff
@@ -1634,7 +1635,7 @@ def makeIteratorRoundTrip(self, backend, file_ending):
             "../samples/unittest_serialIterator." + file_ending,
             io.Access_Type.read_only,
-            jsonConfig
+            jsonConfig,
```
```suggestion
            jsonConfig
```

ax3l commented on test/python/unittest/API/APITest.py:
```diff
@@ -1682,9 +1683,11 @@ def makeAvailableChunksRoundTrip(self, ext):
             name,
-            io.Access_Type.read_only
+            io.Access_Type.read_only,
+            options="{\"defer_iteration_parsing\": true}"
```
Probably not worth here, but for all larger JSON options in Python (e.g. `jsonConfig` above) you can create a pretty Python dictionary and make it JSON like this:
```py
import json

options = {"defer_iteration_parsing": True}
json_str = json.dumps(options)
```

e.g.
```suggestion
            options=json.dumps({"defer_iteration_parsing": True})
```

Simple for here:
```suggestion
            options='{"defer_iteration_parsing": true}'
```

franzpoeschel replied:
Ah, good to know. I'll leave it at this for now, then.

ax3l commented on src/Series.cpp:
```diff
@@ -49,6 +50,20 @@ namespace
      */
     std::string cleanFilename(std::string const &filename, Format f);
 
+    struct Match
```
```suggestion
    /** Compound return type for regex matching of filenames */
    struct Match
```

ax3l commented on src/Series.cpp:
```diff
@@ -49,6 +50,20 @@ namespace
+        uint64_t iteration{};
+
+        // support for std::tie
+        operator std::tuple< bool &, int &, uint64_t & >()
```
Clever :)

ax3l commented on src/Series.cpp:
```diff
@@ -49,6 +50,20 @@ namespace
 
+    struct Match
+    {
+        bool isContained{};
+        int padding{};
+        uint64_t iteration{};
```
```suggestion
        bool isContained{}; //! pattern match successful
        int padding{}; //! number of zeros used for padding of iteration
        uint64_t iteration{}; //! iteration found in regex pattern (default: 0)
```
To avoid confusion, one could make iteration an `int64_t` and return `-1` by default (or so). But we already use `uint64_t` everywhere, so that's fine, too.

franzpoeschel replied:
If we want to do this, I think the cleanest way would be by combining `isContained` and `iteration` into an `auxiliary::Option<uint64_t>`. Should we use something like this right away?

franzpoeschel committed:
Hide parseJsonOptions from public files

This makes all members of SeriesData public, but Series still hides
them.

franzpoeschel committed:
Apply suggestions from code review

Co-authored-by: Axel Huebl <axel.huebl@plasma.ninja>

ax3l committed:
Python: Simplification for options

ax3l reviewed (approved):
Thank you, this is great!
